Feat/pytorch vswa kvcachemanager

qixiang-99 opened the pull request:
# \[Feature] PyTorch VSWA KvCacheManager Support

## Caveat
There is a [WAR](https://github.com/NVIDIA/TensorRT-LLM/pull/5151/files#diff-63c50d73505d372036cf1580e0684c27381bb26b41b49f2ce8713bdf6744d3b3R274-R376) in this PR that is specific for `Gemma3ForCausalLM`, **any model that wants to use VSWA KvCacheManager should modify accordingly.**


### TODOs:

* [x] Expand detailed description.
* [x] Clean up logging.
* [x] Finalize GSM8K benchmark scores comparing HF's and current main's Gemma 3 1b and add it.
* [x] Fix/Update current main Gemma 3 1b test
  * [x] set `enable_block_resue=False`
  * [x] add chunk unit size restriction for chunked prefill correctness 


## Description

### TL;DR
This PR adds:
1. PyTorch runtime support for VSWA KvCacheManager
2. Fix Gemma 3 1b integration test by setting `enable_block_reuse=False` explicitly because current (V)SWA is not compatible with block reuse(`enable_block_reuse=True` by default)
3. Fix Chunked prefill + (V)SWA bug, though this would not be necessary in the future. 

### Background

Previously, Variable Sliding Window Attention (VSWA) support was limited to the C++ KvCacheManager within the C++ runtime. The existing PyTorch implementation only supports a fixed `max_seq_len` window size uniformly across all attention layers. This Pull Request extends PyTorch's KvCacheManager to handle distinct attention window sizes for different layers.

### Implementation Details

To integrate VSWA into PyTorch's KvCacheManager, key changes primarily involve preparing inputs compatible with the underlying C++ Pybind class.

#### Blocks per Window

A significant [update](https://github.com/NVIDIA/TensorRT-LLM/pull/4667) introduces the `blocks_per_window` parameter in the KvCacheManager constructor, which specifies the number of memory blocks in primary and secondary memory for each distinct window size.

Currently, PyTorch's KvCacheManager uses its own implementation (`calculate_max_num_blocks`) that supports a single window size for determining block allocation. After discussion with the team, to ensure consistency between C++ and PyTorch implementations, this PR adds a new wrapper function, `calculate_max_num_blocks_from_cpp`, directly invoking C++ KvCacheManager’s calculation logic.

Notably, we apply the C++ calculation **exclusively** in VSWA contexts, retaining PyTorch's original implementation elsewhere. The rationale for this approach includes:

* Ensuring current PyTorch models remain unaffected.
* Avoiding potential performance regressions across existing models.
* Limiting the complexity of the current PR, deferring broader integration to subsequent detailed evaluations.

Future iterations can progressively transition toward broader adoption of the C++ calculation method following thorough performance analyses.

#### Inputs for C++ `calculate_max_num_blocks`

Given that the C++ `calculate_max_num_blocks` method is a Pybind class member, its inputs require precise types, necessitating conversions in the PyTorch implementation:

##### ModelConfig

PyTorch path utilizes its own `ModelConfig` class extensively. This PR enhances the existing `get_bindings_model_config` function to convert PyTorch’s `ModelConfig` to the necessary `ModelConfigCpp`.

##### ~~KvCacheConfig~~ Solved by https://github.com/NVIDIA/TensorRT-LLM/pull/5384

~~The current PyTorch codebase predominantly passes instances of `tensorrt_llm.bindings.executor.KvCacheConfig` instead of the expected `tensorrt_llm.bindings.KvCacheConfig` to KvCacheManager init function. Although structurally similar, these mismatches lead to type errors when invoking C++ methods. Given the widespread nature of this practice, this PR selectively converts `executor.KvCacheConfig` to `bindings.KvCacheConfig` only as required for compatibility with C++ method calls.~~

### GSM8K Tests for Gemma 3 1b

To verify the functional correctness of PyTorch VSWA KvCacheManager, this PR adds the GSM8K task to the Gemma-3-1b integration tests. To ensure the reference GSM8K scores are representative, GSM8K evaluations were conducted using  Huggingface's implementation of Gemma-3-1b.

Ultimately, an average accuracy of **25.52%** was selected as the reference score. Detailed evaluation steps are documented below. (`avg score = mean(flexible-extract acc, strict-match acc)` )


####  lm_eval with HF Gemma-3-1b GSM8K
This utilizes lm_eval from its [repo](https://github.com/EleutherAI/lm-evaluation-harness).
Run this command:
```bash
lm_eval --model hf \
--model_args pretrained=/home/scratch.trt_llm_data/llm-models/gemma/gemma-3-1b-it/,trust_remote_code=True \
--tasks gsm8k \
--batch_size auto \
--device cuda:0 \
--output_path $OUTPUT_DIR

```

Output:

```bash
2025-06-12:00:00:21,259 INFO     [lm_eval.loggers.evaluation_tracker:209] Saving results aggregated
hf (pretrained=/home/scratch.trt_llm_data/llm-models/gemma/gemma-3-1b-it/,trust_remote_code=True), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: auto
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.2563|±  | 0.012|
|     |       |strict-match    |     5|exact_match|↑  |0.2540|±  | 0.012|

```





## Test Coverage
Use Gemma 3 1b integration test as this is the only VSWA models supported:
`tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestGemma3_1BInstruct`

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->


## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

qixiang-99 committed:
Refactor and pipeclean, can pass gemma3 test. Need to run CI to find out whether there are more issues.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
fix:
- Many tests directly pass executor.KvCacheConfig to KvCacheManager, apply a conversion if this happen
- add `self.blocks_in_primary_pool` and `self.blocks_in_secondary_pool` as these are used for standard case.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
fix: correct the test that passes the kvcache config with the wrong type (the mirror class of executor.kvcache_config)

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
fix:
- relax the type restirction for kv_cache_config and only require as needed.
- update get_cache_indices and get_batch_cache_indices method accordingly
- revert the change of test

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #8627](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8627/) [ run ] triggered by Bot

qixiang-99 committed:
fix:
- update comments
- clean up logging

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
test: Add GSM8K evaluation to Gemma3_1BInstruct tests

- Included GSM8K task evaluation in both test cases for LLM API.
- Updated gsm8k.yaml with accuracy reference for google/gemma-3-1b-it.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #8847](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8847/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6434](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6434/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/6434/test-results/rerun_report.html) for details.

netanel-haber commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -175,6 +176,20 @@ def try_prepare_estimation(self) -> bool:
+                    sliding_window = unique_windows[0] if unique_windows[
+                        0] != self._executor_config.max_seq_len else unique_windows[
+                            1]
+                    self._executor_config.kv_cache_config.max_tokens += (
```
I don't understand this?
The result will be:
`self._get_token_num_for_estimation() + self._executor_config.max_num_tokens + sliding_window`.
What does this calulcation correspond to?

netanel-haber commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -158,46 +162,93 @@ def __init__(
-        self.max_attention_window = max_attention_window if kv_cache_type == CacheTypeCpp.SELF else self.max_seq_len
+        # Set up temp_attention_window_inputs only for VSWA or when needed
+        temp_attention_window_inputs = None
+        if is_vswa:  # Only create when necessary
```
@Funatiq - cc

qixiang-99 replied:
You're right, it would be better to generalize this by creating `temp_attention_window_inputs` for the SWA case rather than limiting it only to VSWA.

Thanks also for pointing out the confusion around `max_num_tokens`. After checking both PyT and C++ implementations, I noticed PyT has a similar "confusion"—there are two separate `max_num_tokens`: one in the runtime config and another in the model config. I'll verify the logic on the PyT sid.

netanel-haber commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -501,6 +565,150 @@ def get_kv_cache_stats(self):
+                raise ValueError(
+                    "max_attention_window_vec cannot be empty if there are local layers."
+                )
+            return {
```
If this is unreachable, raise an exception.

qixiang-99 replied:
Fixed, thanks.

netanel-haber commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -501,6 +565,150 @@ def get_kv_cache_stats(self):
+            }  # Return an empty dict if no local layers or if somehow vec is empty and no layers.
+
+        # Treat max_attention_window_vec as a repeating pattern.
+        # This also correctly handles the case where len(self.max_attention_window_vec) == 1.
```
I think it is more readable if single SWA is treated as a special case + early return, which also guarantees that the default path [the more common case, not VSWA] doesn't get tangled up in the messier logic below

netanel-haber replied:
i.e., 
```python
if len(max_attention_window_vec) == 1:
    return {max_attention_window_vec[0]: list(range(self.num_local_layers))}
```

qixiang-99 replied:
Agreed, modified, thank!

netanel-haber commented on tensorrt_llm/_torch/pyexecutor/resource_manager.py:
```diff
@@ -501,6 +565,150 @@ def get_kv_cache_stats(self):
+            allotted_primary_mem_bytes=primary_pool_memory_bytes,
+            allotted_secondary_mem_bytes=secondary_pool_memory_bytes,
+            extra_cost_memory=extra_cost_memory,
+            kvFactor=self.kv_factor,
```
Not your fault, but I would just fix `kvFactor` to be kv_factor in the bindings at this point

qixiang-99 committed:
fix:
- Explicitly set 'tokens_per_block' for Cpp binding model config. Previously 'tokens_per_block' was default value(64) which might leads to error.
- Remove incorrect "fix" based on misunderstanding that each window size requires additional [max_num_tokens + window_size].

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
fix: address comments

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

Funatiq commented:
Re: `KvCacheConfig` conversion: I'm trying to remove this in #5384.

qixiang-99 commented:
> #5384

Thanks for the update @Funatiq! This change will help improve code consistency. Once you've merged https://github.com/NVIDIA/TensorRT-LLM/pull/5384, I'll update here accordingly.

qixiang-99 committed:
chores: small fix for rebasing, will remove this after #5384 merged

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented on tensorrt_llm/_torch/pyexecutor/_util.py:
```diff
@@ -175,6 +176,20 @@ def try_prepare_estimation(self) -> bool:
+                    # NOTE: For sliding window, we need to allocate more tokens(max_num_tokens + window_size) for each window, need to update if this requirement changes
+                    assert len(
+                        unique_windows
+                    ) == 2, "Sliding window with more than 2 window sizes has not been tested"
```
removed ; )

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #9916](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9916/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #9918](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9918/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #9918](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/9918/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #7321](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/7321/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/7321/test-results/rerun_report.html) for details.

qixiang-99 committed:
when it is swa, we need to make sure chunk size is 256

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
fix: Update the test and GSM8K reference score

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
Remove `KvCacheConfig` conversion as #5384 merged.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
WAR for the kernel diff between Hopper and Blackwell

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #10401](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10401/) [ run ] triggered by Bot

qixiang-99 committed:
fix:  check whether max_attention_window is None

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 committed:
Revert "WAR for the kernel diff between Hopper and Blackwell"
Skip post blackwell for now.

Signed-off-by: qixiang-99 <203170375+qixiang-99@users.noreply.github.com>

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #10438](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/10438/) [ run ] triggered by Bot

qixiang-99 commented:
> Nit: Adjust the PR description for `KvCacheConfig` since it always uses the `executor` one now.

Updated the description, thanks! @Funatiq
